tools/ci_build/reduce_op_kernels.py

```diff
@@ -0,0 +1,256 @@
+# !/usr/bin/env python3
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import argparse
+import op_registration_utils
+import os
+import re
+import shutil
+import sys
+import typing
+
+from logger import get_logger
+
+# add the path to /tools/python so we can import the config parsing and type reduction processing
+script_path = os.path.dirname(os.path.realpath(__file__))
+ort_root = os.path.abspath(os.path.join(script_path, '..', '..', ))
+ort_tools_py_path = os.path.abspath(os.path.join(ort_root, 'tools', 'python'))
+sys.path.append(ort_tools_py_path)
+
+from util import parse_config  # noqa
+from util.ort_format_model.operator_type_usage_processors import OperatorTypeUsageManager  # noqa
+
+log = get_logger("reduce_op_kernels")
+
+
+# valid C++ scalar types that can be specified as globally allowed types
+_valid_allowed_types = {
+    "bool",
+    "uint8_t", "uint16_t", "uint32_t", "uint64_t",
+    "int8_t", "int16_t", "int32_t", "int64_t",
+    "MLFloat16", "BFloat16",  # in onnxruntime namespace
+    "float", "double",
+    "string",  # in std namespace
+}
+
+
+def _validated_globally_allowed_types(globally_allowed_types: typing.Collection[str]) -> typing.Set[str]:
+    '''Return a valid set of globally allowed types.'''
+    # ensure globally_allowed_types is a set
+    if not isinstance(globally_allowed_types, set):
+        globally_allowed_types = set(globally_allowed_types)
+
+    if not globally_allowed_types <= _valid_allowed_types:
+        raise ValueError(
+            "Globally allowed types must be a subset of valid allowed types. Actual: {}, valid: {}".format(
+                globally_allowed_types, sorted(_valid_allowed_types)))
+
+    return globally_allowed_types
+
+
+def _type_re_from_globally_allowed_types(globally_allowed_types: typing.Set[str]) -> typing.re.Pattern:
+    '''Return a regular expression to match type registration strings to a set of globally allowed types.'''
+    # to keep a registration, the type should match patterns like:
+    # 1. T0
+    # 2. T0_T1_T2
+    # where Ti is a member of globally_allowed_types and multiple Ti's are delimited by "_"
+    # this covers both the common case (1) and special cases like OneHot registration (2)
+    allowed_type_subpattern = \
+        "(?:" + "|".join(re.escape(allowed_type) for allowed_type in sorted(globally_allowed_types)) + ")"
+    return re.compile("^{0}(?:_{0})*$".format(allowed_type_subpattern))
+
```

Comment:
Not a big deal as perf isn't a concern here, but would it be simpler to create a set of the types required by the registration (simple split on '_') and check that set is a subset of the globally allowed types?

Reply:
some types can contain underscores, like int8_t, so splitting them would be a bit more complicated.